### ISSUE TITLE
Dequeing code is wrong

### ISSUE THREAD

**Issue by skrysmanski** [SebLague/Procedural-Landmass-Generation#2]
While watching this great tutorial, I noticed that your dequeing code in the `MapGenerator` class is wrong.

First, you're forgetting to lock the queue while dequeing (thus getting potentially undefined behavior).

Secondly, the `for` loop doesn't work as you intended it:

```
for (int i = 0; i < mapDataThreadInfoQueue.Count; i++) {
	MapThreadInfo<MapData> threadInfo = mapDataThreadInfoQueue.Dequeue ();
	threadInfo.callback (threadInfo.parameter);
}
```
The problem here is that `Dequeue()` will reduce `mapDataThreadInfoQueue.Count` by 1 every time it's called. So if you had 2 items in the queue, only the first would be dequeued (because after the first call to `Dequeue()`, `mapDataThreadInfoQueue.Count` would be `1` and so would `i`).

The code should probably be like this:

```
while (mapDataThreadInfoQueue.Count > 0) {
	MapThreadInfo<MapData> threadInfo = mapDataThreadInfoQueue.Dequeue ();
	threadInfo.callback (threadInfo.parameter);
}
```

With this code you can even get rid of the `if (mapDataThreadInfoQueue.Count > 0)` statement.

**Comment by bitinn** [comment]
makes sense, not sure @SebLague is seeing this or not. I was bumping into a similar bug when fiddling with example code.
